drv/stm32xx-gpio-common/src/server.rs

```diff
@@ -0,0 +1,278 @@
+/// GPIO peripherals tend to have many different types, this hides the true type
+/// behind a `&dyn AnyGpioPeriph`.
+///
+/// # Safety
+///
+/// I guess this could be used unsafely if you used it within a system using the
+/// Embedded HAL which relies on uniqueness of peripheral access? In practice,
+/// as long as you ensure you're not doing this, you're probably ok.
+#[cfg(feature = "server-support")]
+pub unsafe fn get_gpio_regs(port: Port) -> &'static dyn AnyGpioPeriph {
+    match port {
+        Port::A => &*device::GPIOA::ptr(),
+        Port::B => &*device::GPIOB::ptr(),
+        Port::C => &*device::GPIOC::ptr(),
+        Port::D => &*device::GPIOD::ptr(),
+
+        #[cfg(feature = "has-port-gpioe")]
+        Port::E => &*device::GPIOE::ptr(),
+        #[cfg(feature = "has-port-gpiof")]
+        Port::F => &*device::GPIOF::ptr(),
+        #[cfg(feature = "has-port-gpiog")]
+        Port::G => &*device::GPIOG::ptr(),
+        #[cfg(feature = "has-port-gpioh")]
+        Port::H => &*device::GPIOH::ptr(),
+        #[cfg(feature = "has-port-gpioi")]
+        Port::I => &*device::GPIOI::ptr(),
+        #[cfg(feature = "has-port-gpioj")]
+        Port::J => &*device::GPIOJ::ptr(),
+        #[cfg(feature = "has-port-gpiok")]
+        Port::K => &*device::GPIOK::ptr(),
+    }
+}
+
+/// A GPIO peripheral that can be manipulated in abstract ways without knowing
+/// its type.
+#[cfg(feature = "server-support")]
+pub trait AnyGpioPeriph {
+    fn configure(&self, pins: u16, atts: u16);
+    fn set_reset(&self, set: u16, reset: u16);
+    fn toggle(&self, pins: u16);
+    fn read(&self) -> u16;
+}
+
+/// Adapter from `GpioPeriph`, the trait implemented for disjoint GPIO
+/// peripheral types, to the general `AnyGpioPeriph`.
+#[cfg(feature = "server-support")]
+impl<T: GpioPeriph> AnyGpioPeriph for T {
+    fn configure(&self, pins: u16, atts: u16) {
+        // The GPIO config registers come in 1, 2, and 4-bit per field variants.
+        // The user-submitted mask is already correct for the 1-bit fields; we
+        // need to expand it into corresponding 2- and 4-bit masks. We use an
+        // outer perfect shuffle operation for this, which interleaves zeroes
+        // from the top 16 bits into the bottom 16.
+
+        // 1 in each targeted 1bit field.
+        let mask_1 = u32::from(pins);
+        // 0b01 in each targeted 2bit field.
+        let lsbs_2 = outer_perfect_shuffle(mask_1);
+        // 0b0001 in each targeted 4bit field for low half.
+        let lsbs_4l = outer_perfect_shuffle(lsbs_2 & 0xFFFF);
+        // Same for high half.
+        let lsbs_4h = outer_perfect_shuffle(lsbs_2 >> 16);
+
+        // Corresponding masks, with 1s in all field bits instead of just the
+        // LSB.  We use multiplication to distribute a constant over all 1
+        // positions because multiplication is cheap on ARMvx-M devices.
+        let mask_2 = lsbs_2.wrapping_mul(0b11);
```

Comment:
Out of curiosity, are we using `wrapping_mul` here to save an overflow check (since overflow shouldn't be possible)?

Reply:
Bingo. Want me to add a comment?

Reply:
Yes please!

Reply:
(I have added a comment)